stan/math/opencl/rev/vari.hpp

```diff
@@ -0,0 +1,190 @@
+ public:
+  /**
+   * The adjoint of this variable, which is the partial derivative
+   * of this variable with respect to the root variable.
+   */
+  T adj_;
+
+  /**
+   * The value of this variable.
+   */
+  T val_;
+
+  /**
+   * Construct a dense matrix_cl variable implementation from a value. The
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_convertible_t<S&, T>* = nullptr>
+  explicit vari_value(S&& x)
+      : chainable_alloc(),
+        adj_(constant(0, x.rows(), x.cols())),
+        val_(std::forward<S>(x)) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+ protected:
+  // to allow access to this constructor from instantinations with different
+  // template parameters
+  template <typename, typename>
+  friend class vari_value;
+  /**
+   * Construct a dense matrix_cl variable implementation from a value and
+   * adjoint.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam R A dense Eigen type that is convertible to `T_val`
+   * @tparam S A dense Eigen type that is convertible to `T_adj`
+   * @param val Value of the constructed variable.
+   * @param adj Adjoint of the constructed variable.
+   */
+  template <typename R, typename S, require_convertible_t<R&, T>* = nullptr,
+            require_convertible_t<S&, T>* = nullptr>
+  vari_value(R&& val, S&& adj)
+      : chainable_alloc(),
+        adj_(std::forward<S>(adj)),
+        val_(std::forward<R>(val)) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+ public:
+  /**
+   * Construct an dense matrix_cl variable implementation from a value. The
+   *  adjoint is initialized to zero and if `stacked` is `false` this vari
+   *  will be not be put on the var_stack. Instead it will only be put on
+   *  a stack to keep track of whether the adjoint needs to be set to zero.
+   *
+   * All constructed variables are added to a stack.  Variables
+   *  should be constructed before variables on which they depend
+   *  to insure proper partial derivative propagation.  During
+   *  derivative propagation, the chain() method of each variable
+   *  will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
```

Comment:
This has to be a `matrix_cl` type right? Idt we have methods to implicitly convert from `Eigen` to `matrix_cl`

Reply:
I think all the `dense Eigen` just needs to be replaced by `matrix_cl`

We could also have a constructor that takes in an eigen type and does `to_matrix_cl`

Reply:
fixed